src/Orleans.Runtime/Placement/Rebalancing/ActiveRebalancerGrain.cs

```diff
@@ -0,0 +1,723 @@
+
+                continue; // there was some problem, try the next best candidate
+            }
+
+            if (response.Type == AcceptExchangeResponse.ResponseType.Success)
+            {
+                await FinalizeProtocol(response.ExchangeSet, exchangeSet.Select(x => x.Id).Union(response.ExchangeSet), isReceiver: false);
+                _currentExchangeSilo = null;
+
+                break;  // exchange was successful, no need to iterate over another candidate.
+            }
+
+            _currentExchangeSilo = null;
+
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                if (response.Type == AcceptExchangeResponse.ResponseType.ExchangedRecently)
+                {
+                    _logger.LogDebug(
+                        "Exchange request from {ThisSilo} failed, due to {OtherSilo} having been recently involved in another exchange. " +
+                        "I will try the next best candidate (if one is available), otherwise I will wait for my next period to come.", ThisSilo, candidateSilo);
+                }
+                else if (response.Type == AcceptExchangeResponse.ResponseType.MutualExchangeAttempt)
+                {
+                    _logger.LogDebug(
+                        "Exchange request from {ThisSilo} failed, due to an mutual exchange attempt with {OtherSilo}. " +
+                        "I will try the next best candidate (if one is available), otherwise I will wait for my next period to come.", ThisSilo, candidateSilo);
+                }
+            }
+        }
+    }
+
+    public async Task<AcceptExchangeResponse> AcceptExchangeRequest(AcceptExchangeRequest request)
+    {
+        if (_currentExchangeSilo is { } silo && request.SendingSilo.IsSameLogicalSilo(silo))
+        {
+            _unblocker?.SetResult();
+            _unblocker = new();
+
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                _logger.LogDebug(
+                    "I got an exchange request from {Silo}, but I am performing one with it at the same time. " +
+                    "I will phase-shift my timer to avoid these conflicts.", request.SendingSilo);
+            }
+
+            // We pick some random time between 'min' and 'max' and than substract from it 'min'. We do this so this silo doesn't have to wait for 'min + random',
+            // as it did the very first time this was started. It is guaranteed that 'random - min' >= 0; as 'random' will be at the least equal to 'min'.
+            RegisterOrUpdateTimer(RandomTimeSpan.Next(_options.MinimumRebalancingDueTime, _options.MaximumRebalancingDueTime) - _options.MinimumRebalancingDueTime);
+
+            return AcceptExchangeResponse.CachedMutualExchangeAttempt;
+        }
+
+        if (_lastExchangeTime is { } time && time.AddTicks(_options.RecoveryPeriod.Ticks) > DateTime.UtcNow)
+        {
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                _logger.LogDebug(
```

Comment:
Should the logging be moved to a (private) Log class and use the logging source generator?
With code and logging-code interleaved it's a bit hard to follow along.

For sure, that could be done in a separate PR too.